openmls/src/treesync/node/leaf_node.rs

```diff
@@ -690,49 +503,103 @@ impl LeafNode {
     }
 }
 
-const LEAF_NODE_SIGNATURE_LABEL: &str = "LeafNodeTBS";
-
-impl Signable for LeafNodeTbs {
-    type SignedOutput = LeafNode;
-
-    fn unsigned_payload(&self) -> Result<Vec<u8>, tls_codec::Error> {
-        self.tls_serialize_detached()
+#[cfg(any(feature = "test-utils", test))]
+impl LeafNode {
+    /// Replace the credential in the KeyPackage.
+    pub(crate) fn set_credential(&mut self, credential: Credential) {
+        self.payload.credential = credential;
     }
+}
 
-    fn label(&self) -> &str {
-        LEAF_NODE_SIGNATURE_LABEL
+impl From<OpenMlsLeafNode> for LeafNode {
+    fn from(leaf: OpenMlsLeafNode) -> Self {
+        leaf.leaf_node
     }
 }
 
-impl SignedStruct<LeafNodeTbs> for LeafNode {
-    fn from_payload(tbs: LeafNodeTbs, signature: Signature) -> Self {
-        Self {
-            payload: tbs.payload,
-            signature,
-        }
-    }
+/// The payload of a [`LeafNode`]
+///
+/// ```text
+/// struct {
+///     HPKEPublicKey encryption_key;
+///     SignaturePublicKey signature_key;
+///     Credential credential;
+///     Capabilities capabilities;
+///
+///     LeafNodeSource leaf_node_source;
+///     select (LeafNode.leaf_node_source) {
+///         case key_package:
+///             Lifetime lifetime;
+///
+///         case update:
+///             struct{};
+///
+///         case commit:
+///             opaque parent_hash<V>;
+///     };
+///
+///     Extension extensions<V>;
+///     ...
+/// } LeafNode;
+/// ```
+#[derive(
+    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserialize, TlsSize,
+)]
+struct LeafNodePayload {
+    encryption_key: EncryptionKey,
+    signature_key: SignaturePublicKey,
+    credential: Credential,
+    capabilities: Capabilities,
+    leaf_node_source: LeafNodeSource,
+    extensions: Extensions,
 }
 
-/// Helper struct to verify incoming leaf nodes.
-/// The [`LeafNode`] doesn't have all the information needed to verify.
-/// In particular is the [`TreeInfoTbs`] missing.
-pub(crate) struct VerifiableLeafNodeTbs<'a> {
-    pub(crate) tbs: &'a LeafNodeTbs,
-    pub(crate) signature: &'a Signature,
+#[derive(
+    Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TlsSerialize, TlsDeserialize, TlsSize,
+)]
+#[repr(u8)]
+pub enum LeafNodeSource {
+    #[tls_codec(discriminant = 1)]
+    KeyPackage(Lifetime),
+    Update,
+    Commit(ParentHash),
 }
 
-impl<'a> Verifiable for VerifiableLeafNodeTbs<'a> {
-    fn unsigned_payload(&self) -> Result<Vec<u8>, tls_codec::Error> {
-        self.tbs.tls_serialize_detached()
-    }
+pub type ParentHash = VLBytes;
 
-    fn signature(&self) -> &Signature {
-        self.signature
-    }
+// -------------------------------------------------------------------------------------------------
```

Comment:
I'm not sure I understand what these are supposed to be.
If there's that much code here that something like a divider is needed, it should go somewhere else and not add even more content (even if it's a comment).

Reply:
It separates `LeafNode`, `LeafNodeTbs`, etc. code. Not sure if we want own modules for this. I can delete them.